core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
    * Get the level we can launch tasks according to delay scheduling, based on current wait time.
    */
   private def getAllowedLocalityLevel(curTime: Long): TaskLocality.TaskLocality = {
-    while (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex) &&
-        currentLocalityIndex < myLocalityLevels.length - 1)
-    {
-      // Jump to the next locality level, and remove our waiting time for the current one since
-      // we don't want to count it again on the next one
-      lastLaunchTime += localityWaits(currentLocalityIndex)
-      currentLocalityIndex += 1
+    // Remove the scheduled or finished tasks lazily
+    def hasNotScheduledTasks(list: ArrayBuffer[Int]): Boolean = {
```

Comment:
Can you give `list` a more descriptive variable name?  Reading top-to-bottom, it's not obvious what it's representing.